core/WASM_core.md

```diff
@@ -0,0 +1,350 @@
+
+Finally, there may be some issues with other things that in native world use syscalls like atomics,
+locks, random number gen, etc. Luckily my experience so far is that those pretty much all just work
+without any, or fairly minor, changes.
+
+Ultimately we have two strategies available to us. We can create interfaces for Core to call back
+into lang to perform all these functions, or we can attempt to leverage WASI for this functionality.
+WASI is quite immature, but *in theory* it can support these use cases.
+
+## WASM Runtime
+
+Regardless of the strategy, we need to choose a WASM runtime which each Lang SDK will use, so I'll
+address that first. The two major players in the space are
+[Wasmer](https://github.com/wasmerio/wasmer) and
+[Wasmtime](https://github.com/bytecodealliance/wasmtime). It's also not impossible that we might
+use one library from Rust for defining things like the host imported functions, but use different
+libraries for the runtime in the host langs.
+
+### Wasmer
+Wasmer's biggest pro is it's language support. Every language we currently have an SDK for (and
+more beyond that) has a Wasmer SDK for embedding and running WASM in that language. It's docs are
+decent, but not great.
+
+It has an MIT license.
+
+It's controlled by a startup (but hey, so are we) and has somewhat less active open-source
+development than Wasmtime.
+
+### Wasmtime
+Wasmtime supports all our languages _except_ Java, which is unfortunate, though it looks like
+there's an unofficial repo. It's also not immediately clear how usable JS/Node support is, 
+particularly when it comes to using WASI. It does have quite excellent docs, and built-in support
+for calling async functions from WASM, which is pretty awesome.
+
+It has an Apache 2.0 license.
+
+It's owned and maintained by the bytecode alliance, which also owns the WASM/WASI spec. It's backed
+by Mozilla.
+
+### Oh no CGo (and JNI)
+There's a big downside to both Wasmer and Wasmtime. Both of their Go implementations use CGo, and
+Wasmer's Java implementation uses JNI. This blows one of the bigger pluses out of the water.
+Luckily, [wazero](https://github.com/tetratelabs/wazero) exists, which we can use to get a pure Go
+environment. It supports WASI as well, but seemingly [not what we would need for network
+calls](https://wazero.io/specs/).
+
+I can't seem to find anything reasonable for Java. Maybe if we're blessed, such a thing
+will exist by the time we get around to porting Java, but that's a huge if.
+
+## Implementation Strategy
+
+### How much, if any WASI to use?
+
+The more we can lean on WASI, the less we have to expand the Core<-->Lang interface. That's a good
+thing in terms of both immediate work and long-term maintenance. There are some problems though:
+
+* Assuming we need to use `wazero` for non-CGo support, it doesn't appear to be able to handle
+  network calls. We'll test this, though.
```

Comment:
Also note that if we want to use WASM Core in the browser we can't make direct gRPC calls and need a grpc-web or grpc-gateway based client.